### ISSUE TITLE
Undeprecate `mp.get_wakeup_pipe()`

### ISSUE THREAD

**Issue by ruihe774** [mpv-player/mpv#14933]
### Expected behavior of the wanted feature

I propose to revert the deprecation of the Lua API `mp.get_wakeup_pipe()` and its underlying C API `mpv_get_wakeup_pipe()`. `mp.get_wakeup_pipe()` is the only method to implement a custom event loop in the Lua script. It is useful when a scripter wants to wait for multiple event sources. (In contrast, `mp.wait_event()` can only listen to events from mpv, not external event sources.)

The deprecation message of `mpv_get_wakeup_pipe()` suggests to use `mpv_set_wakeup_callback()`; however, it is not possible to implement a similar alternative for Lua. The callback set by `mpv_set_wakeup_callback()` can be called in arbitrary threads, and can be called when Lua runtime is concurrently running, which breaks the single-thread assumption of Lua runtime.

**Comment by ruihe774** [comment]
> Not really. You can always do periodic polling. Either by adding a periodic timer with mp.add_periodic_timer() which polls external events, or polling mpv and external events explicitly in a loop.

Yes, but suboptimal.

**Comment by sfan5** [comment]
Isn't the usability of this rubbish?
How do you read from an fd from Lua? Not possible with the stdlib AFAIK.

**Comment by ruihe774** [comment]
> Isn't the usability of this rubbish? How do you read from an fd from Lua? Not possible with the stdlib AFAIK.

Yes, not possible with the stdlib.

However, LuaJIT FFI can be used to call `read`, `poll`, etc, with which it is possible to implement a full-fledged custom event loop (I actually wrote [one](https://github.com/ruihe774/mpv-lua-mpris/blob/main/mpris.lua)); or existing wrappers like [luv](https://github.com/luvit/luv/) can be used, which is easier.

**Comment by ruihe774** [comment]
> it should be possible to create a separate therad in lua which basically implements `mpv_get_wakeup_pipe()`, right?

It's not possible. Creating a separate thread is not allowed in Lua. Lua has no concept about threads. Calling lua functions outside the thread the lua runtime is created will corrupt the runtime.

**Comment by ruihe774** [comment]
> Also if using multiple scripts is an option, the requested feature can be achieved by adding dedicated scripts to listen to mpv and external events and then sending them to the processing script with `script-message-to`.

Sounds possible. But I'm afraid the dedicated script that listens to external events still has to have a way to listen to some fundamental mpv events like `shutdown`, otherwise it cannot even stop when mpv is going to quit.

**Comment by ruihe774** [comment]
FWIW, it is actually possible to create a (named) pipe in Windows to dispatch events, which is what libuv does. I have the idea to make `mpv_get_wakeup_pipe()` cross-platform.

**Comment by ruihe774** [comment]
> The function running in the separate thread can come from FFI which doesn't touch Lua runtime.

Yes, C functions are allowed. But limited things can be done without writing the C function by myself. And consider the signature of the C function must confirm the signature required by `pthread_create` (`void *(*start_routine)(void *)`), I really doubt it is doable.

```lua
ffi.C.pthread_create(
  thread,
  attr,
  ffi.C.thread_routine,
  arg,
)

-- ffi.C.thread_routine must be a function pointer of `void *(*start_routine)(void *)`
-- we have to implement it in C, which is out of the scope of Lua scripting.
```

> the mpv event listening script can send the shutdown notification to one of the fd that external event script is listening.

Sure.
